### PR TITLE
manifest: nrf_modem: v2.5.0 release

### PR DESCRIPTION
**Pull request by eivindj-nordic**
See CHANGELOG file in nrfxlib for details about library features.

https://github.com/nrfconnect/sdk-nrfxlib/pull/1079
https://github.com/nrfconnect/sdk-zephyr/pull/1316

test-sdk-nrf: sdk-nrf-pr-12425

### CODE REVIEW

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 1]
```diff
@@ -460,6 +484,22 @@ static void update_process(void)
 		reboot_work_handler();
 	}
 }
+#endif
+
+static uint8_t apply_firmware_delta_modem_update(void)
```
Comment (jheiskan81) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:489]: Who is calling this static function?

Comment (eivindj-nordic) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c]: Thanks, I've updated with your changes now.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 2]
```diff
@@ -216,6 +214,11 @@ static int firmware_request_linked_update(int instance_id)
 	return postpone_update;
 }
 
+static void on_modem_lib_init(int ret, void *ctx)
```
Comment (jheiskan81) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:217]: This change Lines: 217-221 can be drop it is already done at fota_download_delta_modem.c

**Review by jayteemo — CHANGES_REQUESTED** [review]

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fota_common.c [hunk 3]
```diff
@@ -191,25 +189,28 @@ static enum nrf_cloud_fota_validate_status app_fota_validate_get(void)
...
 		return NRF_CLOUD_FOTA_VALIDATE_FAIL;
 	case NRF_MODEM_DFU_RESULT_VOLTAGE_LOW:
-		LOG_ERR("Modem FOTA cancelled: %d", modem_lib_init_result);
+		LOG_ERR("MODEM UPDATE CANCELLED 0x%x.", dfu_result);
 		LOG_ERR("Please reboot once you have sufficient power for the DFU");
 		return NRF_CLOUD_FOTA_VALIDATE_FAIL;
 	default:
-		LOG_INF("Modem FOTA result unknown: %d", modem_lib_init_result);
+		/* Unknown DFU result code */
+		LOG_WRN("nRF modem DFU result unknown: 0x%x", dfu_result);
```
[the remark this answers is not in the document]
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fota_common.c:210]: missed one 😸 
thanks!

**Review by jayteemo — APPROVED** [review]

**Review by tomi-font — COMMENTED** [review]
Looks good from the SLM's perspective. Just two small points.

File: applications/serial_lte_modem/src/main.c [hunk 4]
```diff
@@ -436,7 +429,7 @@ int main(void)
 
 	const int ret = nrf_modem_lib_init();
 
-	if (ret < 0) {
+	if (ret && ret != -EAGAIN && ret != -EIO) {
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply applications/serial_lte_modem/src/main.c:432]: -EIO is returned on irrecoverable errors (hardware error or internal error), so it won't work, unless it is reprogrammed in bootloader mode or with external tools.

File: applications/serial_lte_modem/src/slm_at_commands.c [hunk 5]
```diff
@@ -216,31 +216,25 @@ static int handle_at_modemreset(enum at_cmd_type type)
...
-		return 0;
-	} while (0);
+	if ((slm_fota_stage == FOTA_STAGE_ACTIVATE
+				&& (slm_fota_type == DFU_TARGET_IMAGE_TYPE_MODEM_DELTA ||
+					slm_fota_type == DFU_TARGET_IMAGE_TYPE_FULL_MODEM))) {
+#if defined(CONFIG_SLM_FULL_FOTA)
+		slm_finish_modem_full_dfu(ret);
+#endif
+		slm_fota_post_process();
+	}
```
Comment (VTPeltoketo) [inline applications/serial_lte_modem/src/slm_at_commands.c:235]: There seems to be some issue somewhere here. In SLM delta modem fota tests, "#XFOTA: 5,0" is missing after AT#XMODEMRESET.

**Review by VTPeltoketo — CHANGES_REQUESTED** [review]

File: applications/serial_lte_modem/src/slm_at_commands.c (hunk 5 above)
Comment (eivindj-nordic) [reply applications/serial_lte_modem/src/slm_at_commands.c:237]: Thank you, yes, as the dfu_handler is called before nrf_modem_lib_init returns, the slm_fota_stage is changed. I've added a fix to get the fota stage before performing the modem reinitialization.

File: applications/asset_tracker_v2/sample.yaml [hunk 6]
```diff
@@ -124,7 +124,6 @@ tests:
     platform_allow: nrf9160dk_nrf9160_ns thingy91_nrf9160_ns
     integration_platforms:
       - nrf9160dk_nrf9160_ns
-      - thingy91_nrf9160_ns
```
Comment (trantanen) [inline applications/asset_tracker_v2/sample.yaml:127]: Is this removal related to nrf_modem 2.5.0 specifically, or just some clean up work?

File: applications/serial_lte_modem/src/slm_at_commands.c (hunk 5 above)
Comment (VTPeltoketo) [reply applications/serial_lte_modem/src/slm_at_commands.c:236]: It seems to work now with XMODEMRESET

Comment (eivindj-nordic) [reply applications/serial_lte_modem/src/slm_at_commands.c:236]: @tomi-font I didn't want to use `FOTA_STAGE_COMPLETE` as I don't know when it is cleared. So if the application were to call XMODEMRESET multiple times, it would send the XFOTA event multiple times, although there were only one DFU.

Comment (eivindj-nordic) [reply applications/serial_lte_modem/src/slm_at_commands.c:236]: @VTPeltoketo Thank you for checking!

File: samples/cellular/modem_shell/src/fota/fota.c [hunk 7]
```diff
@@ -16,58 +16,54 @@
...
-	case 0:
-		mosh_error("FOTA: Expected DFU result from modem library, "
-			   "but no DFU was triggered");
-		goto exit;
+	switch (dfu_res) {
 	case NRF_MODEM_DFU_RESULT_OK:
-		mosh_print("FOTA: Modem firmware update successful!");
-		goto restart_modem;
+		mosh_error("Modem firmware updated");
+		mosh_error("FOTA: Modem firmware update successful!");
```
Comment (trantanen) [inline samples/cellular/modem_shell/src/fota/fota.c:24]: Why two similar prints?
@tokangas

File: samples/cellular/nrf_provisioning/sample.yaml [hunk 8]
```diff
@@ -1,7 +1,7 @@
 sample:
   name: nRF Device Provisioning
 tests:
-  sample.cellular.nrf_provisioning:
+  sample.nrf9160.nrf_provisioning:
```
Comment (trantanen) [inline samples/cellular/nrf_provisioning/sample.yaml:4]: Is this some leftover or why is this renamed back to nrf9160?

**Comment by VTPeltoketo** [comment]
The PR seems to influence to the lwm2m carrier library hanshaking with the server. Previously the handshaking required the up-down-up-down-up manoeuvre in modem, has this changed?

File: applications/asset_tracker_v2/sample.yaml (hunk 6 above)
Comment (eivindj-nordic) [reply applications/asset_tracker_v2/sample.yaml:127]: We are running out of space...
Agreed with @nordic-hani to remove thingy with the carrier overlay for now and re-enable it later if we can reduce the application size.

File: samples/cellular/modem_shell/src/fota/fota.c (hunk 7 above)
Comment (eivindj-nordic) [reply samples/cellular/modem_shell/src/fota/fota.c:24]: I'll remove the first one.

File: samples/cellular/nrf_provisioning/sample.yaml (hunk 8 above)
Comment (eivindj-nordic) [reply samples/cellular/nrf_provisioning/sample.yaml:4]: This is leftover/bad rebase, thanks!

**Comment by eivindj-nordic** [comment]
> The PR seems to influence to the lwm2m carrier library hanshaking with the server. Previously the handshaking required the up-down-up-down-up manoeuvre in modem, has this changed?

Are you talking about during a DFU? 
If so, yes this has changed so there is no longer a need for an extra reboot after the DFU. This will be fixed in the next carrier library release, though it will be broken until then, as agreed with @nordic-hani.

File: applications/asset_tracker_v2/src/main.c [hunk 9]
```diff
@@ -174,41 +174,48 @@ static void sub_state_set(enum sub_state_type new_state)
...
-		/* Initialization successful, no action required. */
-		return;
+static void on_modem_lib_dfu(int dfu_res, void *ctx)
+{
+	switch (dfu_res) {
 	case NRF_MODEM_DFU_RESULT_OK:
-		LOG_DBG("MODEM UPDATE OK. Will run new modem firmware after reboot");
-		break;
+		LOG_DBG("nRF Modem firmware update succeeded");
+		/* Fallthrough */
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply applications/asset_tracker_v2/src/main.c:183]: Good question :/ Updated now.

File: samples/cellular/modem_shell/src/fota/fota.c (hunk 7 above)
Comment (tokangas) [reply samples/cellular/modem_shell/src/fota/fota.c]: Also, the function has changed from `mosh_print()` to `mosh_error()`. It should be `mosh_print()`, because this is not an error.

**Comment by VTPeltoketo** [comment]
> > The PR seems to influence to the lwm2m carrier library hanshaking with the server. Previously the handshaking required the up-down-up-down-up manoeuvre in modem, has this changed?
> 
> Are you talking about during a DFU? If so, yes this has changed so there is no longer a need for an extra reboot after the DFU. This will be fixed in the next carrier library release, though it will be broken until then, as agreed with @nordic-hani.

No, I do not mean DFU but lwm2m carrier library handshake between device and server (for example avsystems.com). The handshaking works with main, but not using this branch.

**Comment by tokangas** [comment]
> No, I do not mean DFU but lwm2m carrier library handshake between device and server (for example avsystems.com). The handshaking works with main, but not using this branch.

Is this related to the LwM2M client utils library and not the LwM2M carrier library? I think the carrier library isn't used with AVSystems?

**Review by tomi-font — APPROVED** [review]

**Review by jheiskan81 — APPROVED** [review]

**Review by SeppoTakalo — APPROVED** [review]

**Review by VTPeltoketo — APPROVED** [review]

**Review by trantanen — APPROVED** [review]

**Review by tokangas — APPROVED** [review]

**Comment by tokangas** [comment]
`iot_positioning` tests are failing because the code tries to use the new nRF Cloud `agnss` REST endpoint, which is not yet available in the production cloud. Releasing of the cloud side changes has been delayed and the current target is 4th of October. 

`[00:00:06.442,047] <err> nrf_cloud_codec_internal: REST error msg: This endpoint is not accessible.`

@moksanen We probably need to temporarily disable the affected tests cases.

**Review by plskeggs — APPROVED** [review]

File: applications/asset_tracker_v2/src/main.c [hunk 10]
```diff
@@ -174,41 +174,48 @@ static void sub_state_set(enum sub_state_type new_state)
...
-	switch (ret) {
-	case 0:
-		/* Initialization successful, no action required. */
-		return;
+static void on_modem_lib_dfu(int dfu_res, void *ctx)
+{
+	switch (dfu_res) {
 	case NRF_MODEM_DFU_RESULT_OK:
-		LOG_DBG("MODEM UPDATE OK. Will run new modem firmware after reboot");
+		LOG_DBG("nRF Modem firmware update succeeded");
```
Comment (lemrey) [inline applications/asset_tracker_v2/src/main.c:182]: I would keep this line the same, just in case there are tests looking for it; the other prints are unchanged too.

Comment (eivindj-nordic) [reply applications/asset_tracker_v2/src/main.c:182]: That would not be correct though, as the modem is already initialized now. But I will change the prefix.

File: subsys/net/lib/download_client/src/download_client.c [hunk 11]
```diff
@@ -223,12 +223,9 @@ static int socket_tls_hostname_set(int fd, const char * const hostname)
...
 {
 	int err;
-	char buf[8] = {0};
 
-	(void) snprintf(buf, sizeof(buf), "pdn%d", pdn_id);
-
-	LOG_INF("Binding to PDN ID: %s", buf);
-	err = setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, &buf, strlen(buf));
+	LOG_INF("Binding to PDN ID: %d", pdn_id);
+	err = setsockopt(fd, SOL_SOCKET, SO_BINDTOPDN, &pdn_id, sizeof(pdn_id));
```
Comment (stig-bjorlykke) [inline subsys/net/lib/download_client/src/download_client.c:228]: The option length must be 4 bytes.

File: lib/bin/lwm2m_carrier/os/lwm2m_carrier.c [hunk 12]
```diff
@@ -14,6 +14,9 @@
 #define LWM2M_CARRIER_THREAD_STACK_SIZE 4096
 #define LWM2M_CARRIER_THREAD_PRIORITY K_LOWEST_APPLICATION_THREAD_PRIO
 
+int nrf_modem_dfu_result;
```
Comment (stig-bjorlykke) [inline lib/bin/lwm2m_carrier/os/lwm2m_carrier.c:17]: `static int`

File: subsys/net/lib/download_client/src/download_client.c [hunk 13]
```diff
@@ -371,6 +369,7 @@ static int client_connect(struct download_client *dl)
 	}
 
 	if (dl->config.pdn_id) {
+		pdn_id = dl->config.pdn_id;
```
Comment (stig-bjorlykke) [inline subsys/net/lib/download_client/src/download_client.c:372]: The tmp variable, which is unused here (!), is not needed.

**Review by stig-bjorlykke — APPROVED** [review]

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 14]
```diff
@@ -664,7 +664,13 @@ Modem libraries
 
 * :ref:`nrf_modem_lib_readme`:
 
-  * Added CEREG event tracking to ``lte_connectivity``.
+  * Added:
+
+    * CEREG event tracking to ``lte_connectivity``.
+    * The :c:macro:`NRF_MODEM_LIB_ON_DFU_RES` macro to add callbacks for modem DFU results.
+
+  * Replaced the use of :c:macro:`SO_BINDTODEVICE` socket option with :c:macro:`SO_BINDTOPDN` to bind the socket to a PDN. The new option takes an integer for the PDN ID instead of a string.
```
Comment (wiba-nordic) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:672]:
```suggestion
  * Replaced the use of :c:macro:`SO_BINDTODEVICE` socket option with :c:macro:`SO_BINDTOPDN` to bind the socket to a PDN.
    The new option takes an integer for the PDN ID instead of a string.
```

**Review by wiba-nordic — CHANGES_REQUESTED** [review]

**Review by wiba-nordic — APPROVED** [review]

File: lib/bin/lwm2m_carrier/os/lwm2m_carrier.c [hunk 15]
```diff
@@ -92,42 +95,51 @@ __weak int lwm2m_carrier_event_handler(const lwm2m_carrier_event_t *event)
...
+	ARG_UNUSED(ctx);
+
+	int result;
+
+	if (nrf_modem_dfu_result) {
+		result = nrf_modem_dfu_result;
+	} else if (ret == -NRF_EPERM) {
 		printk("Modem already initialized\n");
 		return;
-	default:
```
Comment (kacperradoszewski) [inline lib/bin/lwm2m_carrier/os/lwm2m_carrier.c:125]: We might need the `default` back

**Review by lemrey — APPROVED** [review]
Looks good, thanks!

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 16]
```diff
@@ -588,9 +588,13 @@ static void gnss_event_handler(int event)
...
 		on_gnss_evt_fix();
 		break;
-	case NRF_MODEM_GNSS_EVT_AGPS_REQ:
-		LOG_INF("GNSS_EVT_AGPS_REQ");
-		/* This will send A-GNSS/P-GPS requests only if cloud assistance is enabled. */
+	case NRF_MODEM_GNSS_EVT_NMEA:
+		if (IS_ENABLED(CONFIG_SLM_LOG_LEVEL_DBG)) {
+			on_gnss_evt_nmea();
+		}
+		break;
```
Comment (tokangas) [inline applications/serial_lte_modem/src/gnss/slm_at_gnss.c:595]: This code should be deleted. Case `NRF_MODEM_GNSS_EVT_NMEA` has been moved a few lines up.

Comment (eivindj-nordic) [reply applications/serial_lte_modem/src/gnss/slm_at_gnss.c]: Yes, updated now.

File: applications/serial_lte_modem/src/gnss/slm_at_gnss.c [hunk 17]
```diff
@@ -581,16 +581,17 @@ static void gnss_event_handler(int event)
 		on_gnss_evt_pvt();
 		break;
 	case NRF_MODEM_GNSS_EVT_NMEA:
-		on_gnss_evt_nmea();
+		if (IS_ENABLED(CONFIG_SLM_LOG_LEVEL_DBG)) {
+			on_gnss_evt_nmea();
+		}
```
Comment (tokangas) [inline applications/serial_lte_modem/src/gnss/slm_at_gnss.c:586]: This `if` should be removed, it has been replaced by the `#if defined` above.
```suggestion
		on_gnss_evt_nmea();
```

File: west.yml [hunk 18]
```diff
@@ -61,7 +61,7 @@ manifest:
     # https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/guides/modules.html
     - name: zephyr
       repo-path: sdk-zephyr
-      revision: cafad7403fb4eaf5939c73db2fdfecff2bf65b11
+      revision: pull/1316/head
```
Comment (rlubos) [inline west.yml:64]: `c7f3aaa87cf8ab6ca6961ec57f06d65d7c893e5b`

File: west.yml [hunk 19]
```diff
@@ -141,7 +141,7 @@ manifest:
     - name: nrfxlib
       repo-path: sdk-nrfxlib
       path: nrfxlib
-      revision: 7096d02995dc7ea140340aa473ff1754d7257292
+      revision: pull/1079/head
```
Comment (rlubos) [inline west.yml:144]: `50e726589dd622aff6a255335becf6ac92dd32c2`

**Review by rlubos — COMMENTED** [review]
Please update SHAs and rebase

**Review by rlubos — APPROVED** [review]
